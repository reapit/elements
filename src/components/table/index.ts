export * from './table'
export * from './table-body'
export * from './table-cell'
export * from './table-container'
export * from './table-head'
export * from './table-header-cell'
export * from './table-provider'
export * from './table-row'
export * from './table-row-selection'
export * from './table-text'
export * from './table-toolbar'
