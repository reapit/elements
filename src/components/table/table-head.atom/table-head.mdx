import { Meta, Canvas, Controls } from '@storybook/blocks'
import { RenderHtmlMarkup } from '../../../storybook/render-html-markup'
import * as TableHeadStories from './table-head.stories'

<Meta of={TableHeadStories} />

# TableHead

## Basic Usage

<Canvas of={TableHeadStories.BasicUsage} />

<RenderHtmlMarkup of={TableHeadStories.BasicUsage} />
