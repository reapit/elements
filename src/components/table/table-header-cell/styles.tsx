import { styled } from '@linaria/react'
import { CSSProperties } from 'react'

interface TableHeaderCellCSSProperties extends CSSProperties {
  // NOTE: These are public CSS variables that we use to allow CSS-only consumers
  // We do this ourselves instead of leveraging Linaria's dynamic styles because
  // that approach results in randomly-named CSS variables, which would not provide
  // a friendly interface for CSS-only consumers.
  '--tablecell-header-width'?: string
  '--tablecell-header-min-width'?: string
  '--tablecell-header-max-width'?: string
}

interface ElTableHeaderCellProps {
  style?: TableHeaderCellCSSProperties
}

interface ElTableHeaderCellContentProps {
  'data-alignment': string
}

export const ElTableHeaderCell = styled.th<ElTableHeaderCellProps>`
  --tablecell-header-width: auto;
  --tablecell-header-min-width: auto;
  --tablecell-header-max-width: 100%;

  width: var(--tablecell-header-width);
  min-width: var(--tablecell-header-min-width);
  max-width: var(--tablecell-header-max-width);
  gap: var(--spacing-1);
  vertical-align: middle;
  padding: var(--spacing-2);
`
export const ElTableHeaderCellContent = styled.div<ElTableHeaderCellContentProps>`
  width: auto;
  padding: var(--spacing-2) 0px;
  gap: var(--spacing-1);
  color: var(--text-secondary);
  flex: auto;
  align-items: center;
  flex-wrap: wrap;
  display: flex;
  font-family: var(--font-2xs-bold-family, Inter);
  font-size: var(--font-2xs-bold-size);
  font-style: normal;
  font-weight: var(--font-2xs-bold-weight);
  line-height: var(--font-2xs-bold-line_height);
  letter-spacing: var(--font-2xs-bold-letter_spacing);

  &[data-alignment='left'] {
    text-align: left;
    justify-content: start;
  }
  &[data-alignment='center'] {
    text-align: center;
    justify-content: center;
  }
  &[data-alignment='right'] {
    text-align: right;
    justify-content: right;
  }
`
