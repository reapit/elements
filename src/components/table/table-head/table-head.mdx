import { Meta, Canvas, Controls } from '@storybook/blocks'
import { RenderHtmlMarkup } from '../../../storybook/render-html-markup'
import * as TableHeadStories from './table-head.stories'

<Meta of={TableHeadStories} />

# TableHead
This story demonstrates how to use the `TableHead` component in a basic table setup. It shows the table header with 5 columns, including a checkbox, text, and image headers.
The `TableHead` component is used to define the header of the table. It typically includes `TableHeaderCell` elements that represent the table column headers.
You can combine it with other components like `Table`, `TableBody`, and `TableToolbar` to create a fully functional table.

## Basic Usage
<Canvas of={TableHeadStories.BasicUsage} />

## Basic Display With Sticky Header
<Canvas of={TableHeadStories.BasicUsage} />
<Controls />


<RenderHtmlMarkup of={TableHeadStories.BasicUsage} />
