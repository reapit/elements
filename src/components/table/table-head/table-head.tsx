import React, { FC, ReactNode, TableHTMLAttributes } from 'react'
import { ElTableHead } from './styles'

export interface ITableHeadProps extends Omit<TableHTMLAttributes<HTMLTableSectionElement>, 'children'> {
  children?: ReactNode
  isSticky?: boolean
}

export const TableHead: FC<ITableHeadProps> = ({ children, isSticky, ...rest }) => {
  return (
    <ElTableHead role="columnheader" {...rest} data-position={isSticky ? 'sticky' : 'relative'}>
      {children}
    </ElTableHead>
  )
}
