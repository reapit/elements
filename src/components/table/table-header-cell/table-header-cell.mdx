import { Meta, Canvas, Controls } from '@storybook/blocks'
import { RenderHtmlMarkup } from '../../../storybook/render-html-markup'
import * as TableHeaderCellStories from './table-header-cell.stories'

<Meta of={TableHeaderCellStories} />

# TableHeaderCell
Typically TableHeaderCell are used as TableHead children, and its will have `div` children as the wrapper content.

## Basic Usage

<Canvas of={TableHeaderCellStories.BasicUsage} />
<Controls />

<RenderHtmlMarkup of={TableHeaderCellStories.BasicUsage} />
