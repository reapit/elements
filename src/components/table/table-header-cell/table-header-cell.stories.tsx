import { Meta } from '@storybook/react'
import { TableHeaderCell } from './table-header-cell'
import { Table } from '../table/table'
import { TableContainer } from '../table-container'
import { TableToolbar } from '../table-toolbar'
import { TableHead } from '../table-head/table-head'
import { TableBody } from '../table-body'
import { Input } from '../../input'
import { AvatarRectangle } from '../../avatar-rectangle'

const meta: Meta<typeof TableHeaderCell> = {
  title: 'Components/TableHeaderCell',
  component: TableHeaderCell,
}

export default meta

export const BasicUsage = {
  render: (args) => (
    <TableContainer>
      <TableToolbar description="125 Properties" actions="Page Size menu component goes here" />
      <Table>
        <TableHead>
          <TableHeaderCell {...args} maxWidth="30px">
            Checkbox Header
          </TableHeaderCell>
          <TableHeaderCell {...args}>First Header</TableHeaderCell>
          <TableHeaderCell {...args}>Image Header</TableHeaderCell>
          <TableHeaderCell {...args}>Second Header</TableHeaderCell>
          <TableHeaderCell {...args}>Third Header</TableHeaderCell>
        </TableHead>
        <TableBody>
          {Array.from({ length: 10 }, (_, index) => (
            <tr key={index}>
              <td>
                <Input type="checkbox" />
              </td>
              <td>First Column</td>
              <td>
                <AvatarRectangle variant="residential" size="medium" src="https://picsum.photos/seed/picsum/200/300" />
              </td>
              <td>Second Column</td>
              <td>Last Column</td>
            </tr>
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  ),
  args: {
    alignment: 'left',
  },
  argTypes: {
    alignment: {
      control: 'select',
      description: 'Defines label text alignment children',
      options: ['left', 'center', 'right'],
      table: {
        type: { summary: 'enum' },
        defaultValue: { summary: 'left' },
      },
    },
    width: {
      control: 'text',
      description: 'Defines label text width for the children',
    },
    minWidth: {
      control: 'text',
      description: 'Defines label text minWidth for the children',
    },
    maxWidth: {
      control: 'text',
      description: 'Defines label text minWidth for the children',
    },
    flexDirection: {
      control: 'select',
      description: 'Defines TableCell flex direction to row or column without adding new wrapper on child content',
      options: ['column', 'row'],
      table: {
        type: { summary: 'enum' },
        defaultValue: { summary: 'column' },
      },
    },
    isFlexWrap: {
      control: 'boolean',
      description:
        'Defines TableCell wrap or nowrap for the children of we have custom child without adding new wrapper',
    },
  },
}
