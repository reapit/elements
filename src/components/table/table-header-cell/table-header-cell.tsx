import React, { FC, ReactNode, ThHTMLAttributes } from 'react'
import { ElTableHeaderCell, ElTableHeaderCellContent } from './styles'

export interface TableHeaderCellProps extends ThHTMLAttributes<HTMLTableCellElement> {
  children: ReactNode
  alignment?: 'left' | 'center' | 'right'
  width?: string
  minWidth?: string
  maxWidth?: string
}

/**
 * A standard HTML/JSX `<th>` component for use in tables.
 *
 * `TableHeaderCell` is typically used as a child of `TableHead`, providing column headers.
 * It supports customization of alignment, width, and layout properties like flex direction and wrapping behavior.
 *
 * This component is flexible, allowing various styling options to fit different table structures.
 *
 * ### Props (`TableHeaderCellProps`):
 * - `alignment`: Defines text alignment for the content.
 * - `width`: Sets the width of the table header cell.
 * - `minWidth`: Sets the minimum width of the table header cell.
 * - `maxWidth`: Sets the maximum width of the table header cell.
 */

export const TableHeaderCell: FC<TableHeaderCellProps> = ({
  children,
  alignment = 'left',
  width = 'auto',
  maxWidth = '100%',
  minWidth = 'auto',
  style,
  ...rest
}) => {
  return (
    <ElTableHeaderCell
      {...rest}
      data-alignment={alignment}
      style={{
        '--tablecell-header-max-width': maxWidth,
        '--tablecell-header-min-width': minWidth,
        '--tablecell-header-width': width,
        ...style,
      }}
    >
      <ElTableHeaderCellContent data-alignment={alignment}>{children}</ElTableHeaderCellContent>
    </ElTableHeaderCell>
  )
}
