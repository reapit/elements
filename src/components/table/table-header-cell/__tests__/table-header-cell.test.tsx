import { render } from '@testing-library/react'
import { Table } from '../../table/table'
import { TableHead } from '../../table-head/table-head'
import { TableHeaderCell } from '../table-header-cell'

describe('TableHeaderCell', () => {
  test('should match snapshot', () => {
    const { asFragment } = render(
      <Table>
        <TableHead>
          <TableHeaderCell>Table Header Cell</TableHeaderCell>
        </TableHead>
      </Table>,
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
