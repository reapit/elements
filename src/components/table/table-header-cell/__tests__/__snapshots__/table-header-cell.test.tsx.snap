// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TableHeaderCell > should match snapshot 1`] = `
<DocumentFragment>
  <table
    class="mocked-styled-0 el-table"
  >
    <thead
      class="mocked-styled-1 el-table-head"
      data-position="relative"
      role="columnheader"
    >
      <th
        class="mocked-styled-2 el-table-header-cell"
        data-alignment="left"
        role="cell"
        style="--e13cdof9-0: auto; --e13cdof9-1: auto; --e13cdof9-2: auto;"
      >
        <div
          class="mocked-styled-3 el-table-header-cell-content"
          data-flex-direction="column"
          style="--e1qe93km-0: nowrap;"
        >
          Table Header Cell
        </div>
      </th>
    </thead>
  </table>
</DocumentFragment>
`;
